From 04bda9a33137e6c9324a45da0e2b8f6ef25edaa1 Mon Sep 17 00:00:00 2001
From: Yao Zi <me@ziyao.cc>
Date: Mon, 13 Apr 2026 16:36:59 +0000
Subject: [PATCH] riscv: Fix detection when cpuinfo contains no multi-character
 extension

orc_riscv_target_detection_extension() wrongly assumes that there's
at least one underscore in the ISA string, which isn't true on RISC-V
platforms where no CPU core declares any multi-character
extensions, leading to crashes in strcmp().

This isn't reproducible on modern kernels and devicetrees, since
CSR/timer-related instructions (zicsr/zifencei) have been separated out
of RISC-V base extension in later RISC-V specifications, and are
necessary for Linux to boot. But it's still problematic with
vendor-provided older devicetree/kernels, or ones generated by QEMU
earlier than v7.1.0 (without commit 5160bacc0638 ("target/riscv: add
zicsr/zifencei to isa_string")).

Fixes: c9c39f3482fd ("riscv: Add target")
Signed-off-by: Yao Zi <me@ziyao.cc>
---
 orc/riscv/orcriscvtarget.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/orc/riscv/orcriscvtarget.c b/orc/riscv/orcriscvtarget.c
index 503145b67582..259148b1e07b 100644
--- a/orc/riscv/orcriscvtarget.c
+++ b/orc/riscv/orcriscvtarget.c
@@ -60,10 +60,10 @@ orc_riscv_target_detect_extension (const char *exts, const char *ext)
     return found && (found < sep || sep == NULL);
   }
 
-  do {
+  for (; sep; sep = strchr (sep + 1, '_')) {
     if (strncmp (sep + 1, ext, n) && !isalnum (exts[n]))
       return TRUE;
-  } while ((sep = strchr (sep + 1, '_')));
+  }
 
   return FALSE;
 }
-- 
2.53.0

